Tester/CodeCoverage/Generators/AbstractGenerator.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+/**
+ * This file is part of the Nette Tester.
+ * Copyright (c) 2009 David Grudl (http://davidgrudl.com)
+ */
+
+namespace Tester\CodeCoverage\Generators;
+
+
+/**
+ * Code coverage report generator.
+ */
+abstract class AbstractGenerator
+{
+	/** @var array */
+	public $acceptFiles = array('php', 'phpc', 'phpt', 'phtml');
+
+	/** @var array */
+	protected $data;
+
+	/** @var string */
+	protected $source;
+
+
+	/**
+	 * @param  string  path to coverage.dat file
+	 * @param  string  path to covered source file or directory
+	 */
+	public function __construct($file, $source = NULL)
+	{
+		if (!is_file($file)) {
+			throw new \Exception("File '$file' is missing.");
+		}
+
+		$this->data = @unserialize(file_get_contents($file));
+		if (!is_array($this->data)) {
+			throw new \Exception("Content of file '$file' is invalid.");
+		}
+
+		if (!$source) {
+			$source = key($this->data);
+			for ($i = 0; $i < strlen($source); $i++) {
+				foreach ($this->data as $s => $foo) {
+					if (!isset($s[$i]) || $source[$i] !== $s[$i]) {
+						$source = substr($source, 0, $i);
+						break 2;
+					}
+				}
+			}
+			$source = dirname($source . 'x');
+
+		} elseif (!file_exists($source)) {
+			throw new \Exception("File or directory '$source' is missing.");
+		}
+
+		$this->source = realpath($source);
+	}
+
+
+	public function render($file = NULL)
+	{
+		$handle = $file ? @fopen($file, 'w') : STDOUT;
+		if (!$handle) {
+			throw new \Exception("Unable to write to file '$file'.");
+		}
+
+		ob_start(function($buffer) use ($handle) { fwrite($handle, $buffer); }, 4096);
+		try {
+			$this->renderSelf();
+		} catch (\Exception $e) {
+			ftruncate($handle, 0);
```

Comment:
Maybe ob_end_flush can write to file, so what about to move it to `if (isset($e))`? Or close file after ob_end_flush() and unlink it in  `if (isset($e))`?